perf/impl/https/v0.1/main.go

```diff
@@ -21,88 +22,71 @@ import (
 	"time"
 )
 
-const (
-	BlockSize = 64 << 10
-)
+const blockSize = 64 << 10
 
 func handleRequest(w http.ResponseWriter, r *http.Request) {
-	// Read the big-endian bytesToSend value
-	var bytesToSend uint64
-	if err := binary.Read(r.Body, binary.BigEndian, &bytesToSend); err != nil {
-		http.Error(w, "failed to read uint64 value", http.StatusBadRequest)
+	u64Buf := make([]byte, 8)
+	if _, err := io.ReadFull(r.Body, u64Buf); err != nil {
+		log.Printf("reading upload size failed: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
 
-	// Read and discard the remaining bytes in the request
-	io.Copy(io.Discard, r.Body)
-
-	// Set content type and length headers
-	w.Header().Set("Content-Type", "application/octet-stream")
-	w.Header().Set("Content-Length", strconv.FormatUint(bytesToSend, 10))
-
-	// Write the status code explicitly
-	w.WriteHeader(http.StatusOK)
+	bytesToSend := binary.BigEndian.Uint64(u64Buf)
 
-	buf := make([]byte, BlockSize)
+	if _, err := drainStream(r.Body); err != nil {
+		log.Printf("draining stream failed: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
 
-	for bytesToSend > 0 {
-		toSend := buf
-		if bytesToSend < BlockSize {
-			toSend = buf[:bytesToSend]
-		}
+	r.Header.Set("Content-Type", "application/octet-stream")
+	r.Header.Set("Content-Length", strconv.FormatUint(bytesToSend, 10))
 
-		n, err := w.Write(toSend)
-		if err != nil {
-			http.Error(w, "Failed write", http.StatusInternalServerError)
-			return
-		}
-		bytesToSend -= uint64(n)
+	if err := sendBytes(w, bytesToSend); err != nil {
+		log.Printf("sending response failed: %s", err)
+		return
 	}
 }
 
-var zeroSlice = make([]byte, BlockSize) // Pre-made zero-filled slice
-
-type customReader struct {
-	downloadBytes uint64
-	uploadBytes   uint64
-	position      uint64
+type nullReader struct {
+	N    uint64
+	read uint64
 }
 
-func (c *customReader) Read(p []byte) (int, error) {
-	if c.position < 8 {
-		binary.BigEndian.PutUint64(p, c.downloadBytes)
-		c.position += 8
-		return 8, nil
-	} else if c.position-8 < c.uploadBytes {
-		bytesToWrite := min(min(len(p), int(c.uploadBytes-(c.position-8))), len(zeroSlice))
-		copy(p[:bytesToWrite], zeroSlice[:bytesToWrite]) // zero the slice
-		c.position += uint64(bytesToWrite)
-		return bytesToWrite, nil
-	}
-
-	return 0, io.EOF
-}
+var _ io.Reader = &nullReader{}
 
-// Helper function to get minimum of two integers
-func min(a, b int) int {
-	if a < b {
-		return a
+func (r *nullReader) Read(b []byte) (int, error) {
+	remaining := r.N - r.read
+	l := uint64(len(b))
+	if uint64(len(b)) > remaining {
+		l = remaining
 	}
-	return b
+	r.read += l
+	if r.read == r.N {
+		return int(l), io.EOF
+	}
+	return int(l), nil
 }
```

Comment:
I am assuming that it is safe in Golang to return "uninitialized" bytes, i.e. the original bytes in `b` to a remote machine. E.g. there is no risk of `b` containing deallocated secret key material.

Just for others, returning non-zero bytes if just fine according to the perf specification:

> The bytes themselves should be a predetermined arbitrary set of bytes. Zero is fine, but so is random bytes (as long as it's not a different set of random bytes, because then you may be limited by how fast you can generate random bytes).

https://github.com/libp2p/specs/blob/master/perf/perf.md